File: witterungsstation_py/core/runtime_registry.py
from __future__ import annotations

from typing import Any, Callable

from core.registry import GenericRegistry

ServiceFactory = Callable[[Any], Any]
ControllerFactory = Callable[[Any], Any]

_SERVICE_REGISTRY = GenericRegistry[ServiceFactory]()
_CONTROLLER_REGISTRY = GenericRegistry[ControllerFactory]()


def register_service_factory(name: str, factory: ServiceFactory) -> None:
    _SERVICE_REGISTRY.register(str(name), factory)


def register_controller_factory(name: str, factory: ControllerFactory) -> None:
    _CONTROLLER_REGISTRY.register(str(name), factory)


def get_service_factories() -> dict[str, ServiceFactory]:
    return _SERVICE_REGISTRY.snapshot()


def get_controller_factories() -> dict[str, ControllerFactory]:
    return _CONTROLLER_REGISTRY.snapshot()


def validate_runtime_registry(*, required_services: list[str], required_controllers: list[str]) -> None:
    _SERVICE_REGISTRY.validate(required_services, label="service")
    _CONTROLLER_REGISTRY.validate(required_controllers, label="controller")


def reset_runtime_registry() -> None:
    _SERVICE_REGISTRY.clear()
    _CONTROLLER_REGISTRY.clear()

File: frontend/src/core/runtimeRegistry.ts
import type { AppContext } from './context'
import { GenericRegistry } from './registry'

export type ServiceFactory = (ctx: AppContext) => unknown
export type ControllerFactory = (ctx: AppContext) => unknown

const serviceRegistry = new GenericRegistry<ServiceFactory>()
const controllerRegistry = new GenericRegistry<ControllerFactory>()

export function registerServiceFactory(id: string, factory: ServiceFactory): void {
  serviceRegistry.register(String(id), factory)
}

export function registerControllerFactory(id: string, factory: ControllerFactory): void {
  controllerRegistry.register(String(id), factory)
}

export function getServiceFactories(): Record<string, ServiceFactory> {
  return serviceRegistry.snapshot()
}

export function getControllerFactories(): Record<string, ControllerFactory> {
  return controllerRegistry.snapshot()
}

export function validateRuntimeRegistry(options: {
  requiredServices: string[]
  requiredControllers: string[]
}): void {
  serviceRegistry.validate(options.requiredServices, 'service')
  controllerRegistry.validate(options.requiredControllers, 'controller')
}

export function resetRuntimeRegistry(): void {
  serviceRegistry.clear()
  controllerRegistry.clear()
}
